src/Generator/Generators/CSharp/CSharpMarshal.cs

```diff
@@ -124,6 +125,8 @@ public override bool VisitArrayType(ArrayType array, TypeQualifiers quals)
                     if (array.Type.IsPointerToPrimitiveType(PrimitiveType.Void))
                         supportBefore.WriteLineIndent("{0}[i] = new global::System.IntPtr({1}[i]);",
                             value, Context.ReturnVarName);
+                    else if(array.Type.TryGetClass(out @class) && @class.IsRefType)
```

Comment:
A missing space after the "if".
More importantly, I think we may need some desugaring/final pointee/etc on array.Type.